### PR TITLE
John Hill

### PR DESCRIPTION
**Pull request by jphill19**
Completed all the iterations and went above and beyond the scope of expectations by challenging myself.

Key Highlights:

- Strong Typing: Enhanced methods to accept only certain data types, reducing room for error.

- Custom Errors: Implemented custom errors to handle edge cases effectively, with corresponding tests.

Iteration 4 (Option 3):

- Option 3 Completed: Created iteration_4.rb to output analytics as outlined in Option 3.

- FacilityBuilder Adjustments: Modified methods to account for daily hours and holidays closed. Unable to fully test Colorado DMV API as it was unresponsive during testing, therefor I left the colorado method untouched.

- VehicleFactory Enhancements: Added a method to handle ny_registration vehicles, which the data coming from was API was for boats, so this was a bit tricky to tackle. I attempted what I think works best and created  corresponding tests.

I am open to feedback and look forward to it!

Thanks,
John

### CODE REVIEW

File: Iteration_4.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
+require './lib/vehicle_factory'
+require './lib/dmv_data_service'
+require './lib/vehicle.rb'
+
+###### Option 3 ---> Analytics #######
```
Comment (KatBrandt) [inline Iteration_4.rb:5]: Remember, we want to practice OOP which means these methods could be encapsulated in a new class. Maybe a class called Analytics or something, then it should have a corresponding test file to test those methods. :)

File: Iteration_4.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+# Hash for years, make, model
+vehicle_make_count = Hash.new(0)
+vehicle_model_count = Hash.new(0)
+vehicle_years_count = Hash.new(0)
+
+#Hash for counties
+county_count = Hash.new(0)
+
+
+def max_count (data)
```
Comment (KatBrandt) [inline Iteration_4.rb:22]: No space between method name and opening parenthesis

File: Iteration_4.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+puts "\nThe most popular model regisred is: #{max_model[0]}, with a total of #{max_model[1]}/#{factory.vehicles_created.length}\n"
+puts "\nThe most popular make registered is #{max_make[0]}, with a total of #{max_make[1]}/#{factory.vehicles_created.length}\n"
+
+puts "\nVehicle By Years"
+vehicle_years_sorted.each do |key, value|
+  puts "\tYear: #{key}, Total: #{value}"
+end
+
+
+puts "\nMost popular county is #{most_popular_county[0]} with a total of #{most_popular_county[1]}/#{factory.vehicles_created.length}"
```
Comment (KatBrandt) [inline Iteration_4.rb:66]: Overall, great job with these analytics! You got lots of good practice with enumerables and hashes.

File: lib/facility.rb [hunk 4]
```diff
@@ -1,14 +1,142 @@
...
+  end
+end
+
+class VehicleAlreadyRegisted < StandardError
+  attr_reader :message
+  def initialize(msg="This vehicle is already registered")
+    @message = msg
+    super(@message)
+  end
+end
```
Comment (KatBrandt) [inline lib/facility.rb:26]: I like this, but I think it could help be more reusable if we had one CustomError class and then it can take in an argument of the message since the structure of these classes is the same. Regardless, great job playing with this!!!!

File: lib/facility.rb [hunk 5]
```diff
@@ -1,14 +1,142 @@
...
+        @collected_fees += 200
+
+      elsif vehicle.antique?
+        vehicle.set_platetype(:antique)
+        @collected_fees += 25
+
+      else
+        vehicle.set_platetype(:regular)
+        @collected_fees += 100
+
```
Comment (KatBrandt) [inline lib/facility.rb:70]: Too many white spaces here, remove white space on line 50, 58, 62, 66, and 70. You _could_ put a white space before an if block though, so a space above line 55.

File: lib/facility.rb [hunk 6]
```diff
@@ -1,14 +1,142 @@
...
+      vehicle.set_registration_date
+
+      return @registered_vehicles << vehicle
+
+      # return "#{vehicle.model} has been registered"
+
+    else
+      return @registered_vehicles
+    end
+
```
Comment (KatBrandt) [inline lib/facility.rb:82]: Good job with the logic here, but this method is doing a lot. How could we refactor this so that we follow SRP - Single Responsiblility Principle and this method is doing one thing? Could be that one thing is orchestrating the flow of data by calling other methods that are holding this logic. More helper methods. :)

File: lib/facility.rb [hunk 7]
```diff
@@ -1,14 +1,142 @@
...
+        vehicle.set_platetype(:antique)
+        @collected_fees += 25
+
+      else
+        vehicle.set_platetype(:regular)
+        @collected_fees += 100
+
+      end
+
+      vehicle.set_registration_date
```
Comment (KatBrandt) [inline lib/facility.rb:73]: I like that you put the logic for setting the reg date in the Vehicle class. The class that owns the attribute should be the one changing that attribute.

File: lib/facility.rb [hunk 8]
```diff
@@ -1,14 +1,142 @@
...
+    else
+      return @registered_vehicles
+    end
+
+  end
+
+  def administer_written_test(registrant)
+    if @services.include?('Written Test')
+      raise TypeError, "registrant must be a registrant object" unless registrant.is_a?(Registrant)
+      raise WrittenTestAlreadyTakenError.new() unless registrant.license_data[:written] == false
```
Comment (KatBrandt) [inline lib/facility.rb:88]: In Ruby, we do not use open parenthesis to invoke a method, so just `.new` not `.new()`

File: spec/facility_spec.rb [hunk 9]
```diff
@@ -2,25 +2,258 @@
...
+      @facility_1.add_service('New Drivers License')
+      @facility_1.add_service('Renew Drivers License')
+      @facility_1.add_service('Vehicle Registration')
+      expect(@facility_1.services).to eq(['New Drivers License', 'Renew Drivers License', 'Vehicle Registration'])
+    end
+  end
+
+
+  describe '#register_vehicle' do
+
```
Comment (KatBrandt) [inline spec/facility_spec.rb:39]: Do not use white space between a describe and an it, or on line 45 between the last line and the end.

File: spec/facility_spec.rb [hunk 10]
```diff
@@ -2,25 +2,258 @@
...
+    it "will only function if facility has the service to do so" do
+      expect(@facility_1.register_vehicle(@cruz)).to eq([])
+
+      @facility_1.add_service('Vehicle Registration')
+      expect(@facility_1.register_vehicle(@cruz)).to eq([@cruz])
```
Comment (KatBrandt) [inline spec/facility_spec.rb:44]: Good job testing the initial state, running the command method, then showing the next value of the attribute.

File: spec/facility_spec.rb [hunk 11]
```diff
@@ -2,25 +2,258 @@
...
+
+      expect(@facility_1.collected_fees).to eq(25)
+      expect(@facility_1.registered_vehicles).to eq([@camaro])
+      expect(@camaro.registration_date).to be_an_instance_of(Date)
+      expect(@camaro.plate_type).to eq(:antique)
+
+
     end
+
+    it 'can appropriately register regular vehicle' do
```
Comment (KatBrandt) [inline spec/facility_spec.rb:86]: Good job testing each of the different kinds of vehicles.

File: spec/facility_spec.rb [hunk 12]
```diff
@@ -2,25 +2,258 @@
...
+      expect(@facility_1.renew_drivers_license(@registrant_1)).to eq(true)
+
+      # Registrant now has renewed_license
+      expect(@registrant_1.license_data).to eq({:written=>true, :license=>true, :renewed=>true})
+
+    end
+    
+    # No edge case test for renewed_license as the assumption is that a user with a renewed_license can renew again
+
+  end
```
Comment (KatBrandt) [inline spec/facility_spec.rb:255]: Good job testing all the conditions for these methods!!!!!

File: lib/facility_builder.rb [hunk 13]
```diff
@@ -0,0 +1,92 @@
+class FacilityBuilder
+
```
Comment (KatBrandt) [inline lib/facility_builder.rb:2]: No space on line 2, but DO want a space between the attr_readers and the def initialize

File: lib/facility_builder.rb [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+  attr_reader :co_facilities, :ny_facilities, :mo_facilities
+  def initialize
+    @co_facilities = []
+    @ny_facilities = []
+    @mo_facilities = []
+  end
+
+
+  def create_co_facilities(data)
+
```
Comment (KatBrandt) [inline lib/facility_builder.rb:12]: No white space here

File: lib/facility_builder.rb [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+      @co_facilities << new_facility
+
+    end
+    return @co_facilities
+  end
+
+  def create_ny_facilities(data)
+
+    @ny_facilities = data.map do |data|
```
Comment (KatBrandt) [inline lib/facility_builder.rb:38]: You used a .each for co, but a .map for ny. Probably just the .each would work for both since you're storing them in the instance variable

File: lib/facility_builder.rb [hunk 16]
```diff
@@ -0,0 +1,92 @@
+class FacilityBuilder
+
+  attr_reader :co_facilities, :ny_facilities, :mo_facilities
+  def initialize
+    @co_facilities = []
+    @ny_facilities = []
+    @mo_facilities = []
```
Comment (KatBrandt) [inline lib/facility_builder.rb:7]: I like your choice of storing these as instance variables.

File: lib/registrant.rb [hunk 17]
```diff
@@ -0,0 +1,32 @@
+#type error link: https://www.exceptionalcreatures.com/bestiary/TypeError.html
+
+class Registrant
+  attr_accessor :license_data
+  attr_reader :name, :age, :permit
```
Comment (KatBrandt) [inline lib/registrant.rb:5]: Careful here, you have exposed a method of .permit that reads back the value of the @permit attribute, but you have another method .permit? that does that same thing. We only want one, the .permit? one so we don't want .permit in the attr_reader.

File: lib/registrant.rb [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+  def permit?
+    @permit
+  end
+
+  def earn_permit
+    if @permit
+      return @permit
+    else @permit
+      @permit = true
+    end
```
Comment (KatBrandt) [inline lib/registrant.rb:28]: If we just do `@permit = true` it will accomplish all this behavior, no need for an if statement here.

File: lib/vehicle.rb [hunk 19]
```diff
@@ -1,18 +1,22 @@
 require 'date'
 
 class Vehicle
+  attr_accessor :registration_date
```
Comment (KatBrandt) [inline lib/vehicle.rb:4]: No need for an attr_accessor here since you created a method of `set_registration_date` that handles changing this attribute.

File: lib/vehicle.rb [hunk 20]
```diff
@@ -22,4 +26,13 @@ def antique?
...
     @engine == :ev
   end
+
+  def set_platetype(type)
+    raise TypeError, "plate_type must be a symbol" unless type.is_a?(Symbol)
+    @plate_type = type
+  end
+
+  def set_registration_date
+    @registration_date = Date.new(Date.today.year, Date.today.month, Date.today.day)
```
Comment (KatBrandt) [inline lib/vehicle.rb:36]: Also could just do `@registration_date = Date.today` :) Nice job making a method for this attribute change!

**Review by KatBrandt — COMMENTED** [review]
Overall, really great job! Nice work challenging yourself with extensions. It looks like you have a good grasp on enumerables and hashes. Testing looks good overall. Let's work on reducing white spaces (helps with readability) and work on SRP - using helper methods to ensure each method is doing one thing.
